dispass.1
.\" Man page generated from reStructuredText.
.
.TH "DISPASS" "1" "March 18, 2016" "0.3" "DisPass"
.SH NAME
dispass \- DisPass Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
dispass [\-f <labelfile>] [\-h] [\-V] <command> [<args>]
.sp
gdispass
.SH SUMMARY
.sp
DisPass is a password manager for GNU/Linux, *BSD, MacOS X and Windows.
It can be used as any traditional password manager, but has one key
difference. DisPass does not store your passwords anywhere, so you
can never lose them. It creates strong and unique passphrases formed
from a master password and a label (and some optional parameters),
helping you get rid of the bad habit of using a single password for
multiple websites. Dispass is a console application, but also has a
simple graphical interface.
.SH DOCUMENTATION
.SS Resources
.INDENT 0.0
.IP \(bu 2
Main website: \fI\%https://dispass.org/\fP
.IP \(bu 2
Python Package Index: \fI\%http://pypi.python.org/pypi/DisPass/\fP
.IP \(bu 2
Agile project management: \fI\%https://waffle.io/babab/DisPass\fP
.IP \(bu 2
Github: \fI\%https://github.com/babab/dispass/\fP
.IP \(bu 2
Bitbucket: \fI\%https://bitbucket.org/babab/dispass/\fP
.UNINDENT
.SS Definitions
.sp
Here are some definitions which may help you understand the rest of the
documentation better.
.INDENT 0.0
.TP
.B label
A label is a string that you use to identify the passphrase. This
can be a domain name of the service the passphrase is used for,
e.g. \(aqgoogle.com\(aq.
.UNINDENT
.sp
Since this program asks for a password/passphrase to generate another
password/passphrase, things may get a bit confusing. I\(aqve decided to use the
words \(aqpassword\(aq and \(aqpassphrase\(aq differently and consistent.
.INDENT 0.0
.TP
.B password
Use of the word \(aqpassword\(aq is dedicated to the input password, i.e. the
password you are asked to enter and only you know. Whenever you read
password, this is the one I mean. The password needs to be at least 8
characters long and does not have a maximum length.
.TP
.B passphrase
With \(aqpassphrase\(aq I always mean the output passphrase, i.e. the passphrase
that is unique and generated from a label, password and sequence number.
Generated passphrases are 30 characters long. The length can be optionally
changed.
.UNINDENT
.SS Labelfile
.sp
DisPass does not store your passwords or passphrases, but it does store
labels and their settings (algorithm, length and sequence_number) for
convenience. However, please keep in mind that you should try to name
labels in a way that you can easily \(aqdispass generate\(aq a passphrase on
any computer/device that has DisPass at any given moment without having
your labelfile at hand.
.sp
When dispass is run without arguments it will try to find a labelfile.
The location of this file varies and depends on the platform type you use,
the file flag and the environment variables that may be set:
.sp
You can override the location of the labelfile using the
\fB\-f <labelfile>, \-\-file=<labelfile>\fP flag. This can be a way for you
to use different sets of labels/passphrases with a different \(aqmaster\(aq
password for each set.
.INDENT 0.0
.IP 1. 3
If \-f flag is given, that value is used.
.IP 2. 3
If environment var DISPASS_LABELFILE is set, that value is used.
.IP 3. 3
If environment var XDG_DATA_HOME is set,
\fB$XDG_DATA_HOME/dispass/labels\fP is used.
.IP 4. 3
If none of the above applies, the labelfile will default to the following
locations:
.INDENT 3.0
.IP \(bu 2
\fBGNU/Linux and Mac OS X\fP: \fB~/.dispass/labels\fP
.IP \(bu 2
\fB*BSD and other Unixen\fP: \fB~/.dispass/labels\fP
.IP \(bu 2
\fBWindows\fP:   \fBC:\eUsers\e<username>\edispass\elabels\fP
.UNINDENT
.UNINDENT
.SS Basic usage
.sp
You can start using dispass for e.g. google.com like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass generate google.com
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The passphrases created are 30 characters long by default, but some
websites may not validate such a long passphrase or you might want to
make it even longer. You can easily set a desired passphrase length
using the \fB\-l\fP flag. Hotmail passwords are limited to 16 characters:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass generate \-l 16 hotmail
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Generating passphrases for multiple labels is just as easy:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass generate google hotmail YaHo0 "P0551bly*a81t)H4rd2rmbr"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Labels are case\-sensitive and digits and special characters can be used.
You should try to name labels in a way that you can easily generate a
passphrase on any computer/device that has DisPass at any given moment.
.SS Label specifications
.sp
While the above works fine without storing any information, you are
encouraged to store your labels in a labelfile for convenience. That
way, you don\(aqt have to use the \fB\-l 16\fP option everytime you create
your hotmail password.
.sp
Labels have a specification that consists of the following parameters:
.INDENT 0.0
.IP \(bu 2
length (the length of the passphrase)
.IP \(bu 2
algorithm (the algorithm to use)
.IP \(bu 2
sequence number (not used in default algorithm)
.UNINDENT
.sp
The \fBlabelspec\fP looks like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
label[:size[:algorithm[:sequence_number]]]
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Adding a label
.sp
You can save a new \fBhotmail\fP label with a length of 16 and generate the
passphrase in one go:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass add \-g hotmail:16
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The next time you generate a passphrase for \fBhotmail\fP using:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass generate hotmail
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
it will return the same passphrase as before (with a length of 16
characters).
.SS Incrementing sequence numbers
.sp
You might want to change the passphrase you enter into some system.
Of course, you can do this simply by using a different label, e.g.:
\fBHotmail\fP or \fBhotmail\-2\fP\&. You don\(aqt have to though.
.sp
DisPass supports dealing with this situation in the \fIdispass2\fP algorithm.
It basically gives you an option of \fIbumping\fP a passphrase by simply
incrementing a \fBsequence number\fP\&.
.sp
To use sequence numbers with the hotmail label from before you can
update it to use the dispass2 algorithm with sequence number 1:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass update hotmail 16:dispass2:1
dispass generate hotmail
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The next time you want to use a different passphrase, you can simply use
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
dispass increment hotmail
dispass generate hotmail
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Using gdispass
.sp
You can start using gDisPass by running the \fBgdispass\fP executable.
Fill in a name for the label that you can easily remember.
.sp
To generate a passphrase for a new label, i.e. a label that you have
never used before, check the appropriate checkbox. This will allow you
to enter the (master) password twice. gDisPass will then compare the
passwords to see if they are the same. This is needed to minimize the
risk of typos. It is advised that you check the box everytime you
create a passphrase for a new label.
.sp
Subsequential generation of passphrases for the same label most probably do
not need this check. You will likely be warned when/if you made a typo by
the system or website you want to authenticate for.
.sp
If you correctly entered a label and password, you can generate the
passphrase by pressing <Return> or by clicking the appropriate button.
The resulting passphrase will be focused and selected. On platforms
that support it (e.g. *BSD or GNU/Linux) the passphrase will be
automatically placed into your copy/paste buffer.
.sp
Resetting all fields when you are done or when you need to quickly
cancel the generation (because someone is watching over your
shoulders) can be done by pressing <Escape> or by clicking the
appropriate button.
.SH OPTIONS
.SS dispass
.INDENT 0.0
.TP
.BI \-f \ <labelfile>\fP,\fB \ \-\-file\fB= <labelfile>
override labelfile
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-V\fP,\fB  \-\-version
show full version information
.UNINDENT
.SS dispass add
.sp
Add a new label to the labelfile and generate passphrase.
The labelspec looks like this: \fBlabel[:size[:algorithm[:sequence_number]]]\fP
.sp
dispass add [\-g] [\-n] [\-s] <labelspec> [<labelspec2>] [...]
.sp
dispass add [\-i] [\-g] [\-h]
.INDENT 0.0
.TP
.B \-i\fP,\fB  \-\-interactive
add label in an interactive manner
.TP
.B \-g\fP,\fB  \-\-generate
immediately generate passphrase after adding it
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-dry\-run
do not actually add label to labelfile
.TP
.B \-s\fP,\fB  \-\-silent
do not print success message
.UNINDENT
.SS dispass disable
.sp
Disable a label without throwing it away
.sp
dispass disable <label>
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-dry\-run
do not actually update label in labelfile
.TP
.B \-s\fP,\fB  \-\-silent
do not print success message
.UNINDENT
.SS dispass enable
.sp
Enable a label
.sp
dispass enable <label>
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-dry\-run
do not actually update label in labelfile
.TP
.B \-s\fP,\fB  \-\-silent
do not print success message
.UNINDENT
.SS dispass generate
.sp
Generate passphrases for one or more labels
.sp
Use the \fB\-v\fP flag to ask for password twice to avoid typing errors
.sp
dispass generate [options] <label> [<label2>] [<label3>] [...]
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-v\fP,\fB  \-\-verify
verify password
.TP
.BI \-l \ <length>\fP,\fB \ \-\-length\fB= <length>
length of passphrase
.TP
.BI \-a \ <algorithm>\fP,\fB \ \-\-algo\fB= <algorithm>
algorithm to use for generation
.TP
.BI \-s \ <seqno>\fP,\fB \ \-\-seqno\fB= <seqno>
sequence number to use for generation
.TP
.BI \-p \ <password>\fP,\fB \ \-\-password\fB= <password>
password to use for generation
.TP
.B \-o\fP,\fB  \-\-stdout
output passphrase(s) directly to stdout
.TP
.B \-\-silent
do not show a prompt when errors occur
.UNINDENT
.SS dispass gui
.sp
Start the graphical version of DisPass.
.sp
dispass gui [\-h]
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.UNINDENT
.SS dispass help
.sp
Show help information
.sp
dispass help [<command>]
.SS dispass increment
.sp
Increment the sequence number of a label
.sp
dispass increment [\-n] [\-s] <label>
.sp
dispass increment [\-h]
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-dry\-run
do not actually update label in labelfile
.TP
.B \-s\fP,\fB  \-\-silent
do not print success message
.UNINDENT
.SS dispass list
.sp
Print a formatted table of labelfile contents
.sp
If \fB\-\-script\fP is passed the output will be optimized for easy
parsing by other programs and scripts by not printing the header
and always printing one entry on a single line using the
following positions:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
Column  1\-50: labelname        50 chars wide
Column 52\-54: length            3 chars wide
Column 56\-70: hash algo        15 chars wide
Column 72\-74: sequence number   3 chars wide
Column 76\-77: disabled          1 char wide
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
dispass list [\-h] [\-n] [\-\-script]
.INDENT 0.0
.TP
.B \-a\fP,\fB  \-\-all
include disabled labels
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-names\-only
only print names of the labels
.TP
.B \-\-script
output in fixed columns
.UNINDENT
.SS dispass remove
.sp
Remove label from labelfile
.sp
dispass remove [\-n] [\-s] <labelname> [<labelname2>] [...]
.sp
dispass remove [\-i] [\-h]
.INDENT 0.0
.TP
.B \-i\fP,\fB  \-\-interactive
remove label in an interactive manner
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-dry\-run
do not actually remove label from labelfile
.TP
.B \-s\fP,\fB  \-\-silent
do not print success message
.UNINDENT
.SS dispass update
.sp
Update information for a label
.sp
dispass update [\-n] [\-s] <label> [<size>]:[<algorithm>]:[<sequence_number>]
.sp
dispass update [\-h]
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help information
.TP
.B \-n\fP,\fB  \-\-dry\-run
do not actually update label in labelfile
.TP
.B \-s\fP,\fB  \-\-silent
do not print success message
.UNINDENT
.SS dispass version
.sp
Show full version information
.sp
dispass version
.SH ACKNOWLEDGEMENTS
.sp
Many thanks go out to Tom (ryuslash) Willemsen for valuable contributions to
gdispass and the new algorithm. He also wrote an awesome wrapper for Emacs so
you can use DisPass in your favorite editor.
.SH SEE ALSO
.INDENT 0.0
.TP
.B Main website with full documentation
\fI\%http://dispass.babab.nl\fP
.TP
.B The cheeseshop (PyPI) project page
\fI\%http://pypi.python.org/pypi/DisPass/\fP
.TP
.B Github repository and Issue tracker
\fI\%https://github.com/dispass/dispass/\fP
.TP
.B IRC
#dispass at Freenode (chat.freenode.net)
.TP
.B Emacs wrapper
\fI\%http://ryuslash.org/projects/dispass.el.html\fP
.UNINDENT
.SH AUTHOR
Benjamin Althues
.SH COPYRIGHT
2011-2016  Benjamin Althues & Tom Willemse
.\" Generated by docutils manpage writer.
.


ChangeLog
ChangeLog
==============================================================================

**Development version** currently in Github / Bitbucket

 * Make DisPass run from Python shell without exiting
 * Remove dispass.el emacs wrapper, maintained separately by Tom Willemsen

**v0.1-alpha-8**  released June 21st, 2012

 * Add dispass.el emacs wrapper, authored by Tom Willemsen (ryuslash)
 * Allow generating a list of passphrases with varying lengths
 * Add labelfile handler and skel/dot.dispass
 * Add -f <labelfile>, --file=<labelfile> option
 * Use labelfile at $HOME/.dispass if no labels are specified
 * Add -l <length> --length= option
 * Do not autostart gui on Windows
 * Explicitly name the gui version 'gDisPass'
 * Autofill 1st column on output via stdout
 * Remove platform name from usage/help
 * Add manpage
 * Use a landing page for the html documentation


**v0.1-alpha-7**  released May 28th, 2012

 * Distribute as package instead of a single module
 * Add LICENSE file
 * Rename README to README.rst
 * Add 'master' Makefile for building documentation


**v0.1-alpha-6**  released May 24th, 2012

 * Initial release of Dispass as single python module



dispass.py
# vim: set et ts=4 sw=4 sts=4:

'''Generate and disperse/dispell passwords'''

# Copyright (c) 2011-2012 Benjamin Althues <benjamin@babab.nl>
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

__docformat__ = 'restructuredtext'

__author__ = "Benjamin Althues"
__copyright__ = "Copyright (C) 2011-2012 Benjamin Althues"
__version_info__ = (0, 1, 0, 'alpha', 6)
__version__ = '0.1a6'

versionStr = 'DisPass ' + __version__

class globalSettings:
    '''Global settings used in controlling program flow'''
    useCurses = None
    '''Switch passphrase output to stdout if not True'''

    hasTk = None
    '''False if Tkinter could not be imported'''

settings = globalSettings()

# Python stdlib - required
import base64
import getopt
import getpass
import hashlib
import os
import re
import sys

# Python stdlib - optional
try:
    import curses
    settings.useCurses = True
except ImportError:
    settings.useCurses = False
try:
    from Tkinter import *
    import tkMessageBox
    settings.hasTk = True
except ImportError:
    settings.hasTk = False

class GUI:
    '''Houses all GUI related objects and interactions'''

    font = "Verdana"
    '''Default font (Verdana)'''

    fontsize = 10
    '''Default fontsize (10 pt.)'''

    def __init__(self):
        '''Initialize GUI object, create the widgets and start mainloop'''

        # Check if Tkinter has been loaded succesfully; else exit
        if not settings.hasTk:
            print 'Could not find Tkinter, this is a package needed '\
                    'for using\nthe graphical version of dispass.'
            print 'To install, search for a python-tk package for your OS.\n'
            print 'Debian / Ubuntu\t\t$ sudo apt-get install python-tk'
            print 'OpenBSD        \t\t# pkg_add -i python-tk'
            return

        self.root = Tk()
        self.root.title(versionStr)
        self.createWidgets(self.root)
        self.root.mainloop()

# GUI # Setters and getters
    def setFont(self):
        '''Set font and fontsize'''
        pass

    def getFont(self, sizediff=0):
        '''Get `font` and `fontsize`, optionally differ from default `fontsize`

        :Parameters:
            - `sizediff`: The difference in pt. from the default `fontsize`

        :Return:
            - Tuple of (`font`, `fontsize`) to be used when creating widgets
        '''
        return (self.font, self.fontsize + sizediff)

# GUI # Prototypes
    def warn(self, message, warning_type='soft', box_title=''):
        '''Prototype for warning user

         * soft warnings display a message in the passwordout field
         * hard warnings do the same and also display a messagebox

        :Parameters:
            - `message`: The message string for warning the user
            - `warning_type`: Either 'soft' (default value) or 'hard'
            - `box_title`: Optional title for tkMessageBox on hard warnings
        '''
        if warning_type == 'soft' or warning_type == 'hard':
            self.result.config(fg="black", readonlybackground="red")
            self.passwordout.set('- ' + message + ' -')

        if warning_type == 'hard':
            self.passwordin1.delete(0, END)
            self.passwordin2.delete(0, END)
            tkMessageBox.showwarning(box_title, message)

# GUI # Event actions
    def OnGen(self):
        '''Check user input

        Warn when user input is insufficient or wrong. Create digest and
        display the generated password if user input is OK.
        '''

        label = self.label.get()
        passwordin1 = self.passwordin1.get()
        passwordin2 = self.passwordin2.get()
        isnew = self.isnew.get()

        if len(label) == 0:
            self.warn('No password generated, label field is empty')
            return
        elif len(passwordin1) == 0:
            self.warn('No password generated, password field is empty')
            return
        elif len(passwordin1) < 8:
            self.warn('Password must contain at least 8 characters', 'hard',
                    box_title='Password is too short')
            return
        elif isnew and passwordin1 != passwordin2:
            self.warn('Passwords are not identical, please try again', 'hard',
                    box_title='Password mismatch')
            return

        # All checks passed, create digest
        h = digest(label + passwordin1)
        self.result.config(fg="black", readonlybackground="green")
        self.passwordout.set(h)

    def OnNew(self):
        '''Toggle double checking of input password'''
        if self.isnew.get() == 0:
            # Disable double check (default)
            self.passwordin2.delete(0, END)
            self.passwordin2.config(state=DISABLED)
        else:
            # Password is new, allow for double checking passwordin
            self.passwordin2.config(state=NORMAL)

    def OnClear(self):
        '''Clear all input fields'''
        self.label.delete(0, END)
        self.passwordin1.delete(0, END)
        self.passwordin2.delete(0, END)
        self.passwordout.set('- No password generated -')
        self.result.config(fg="black", readonlybackground="gray")

# GUI # Create Widgets
    def createWidgets(self, master):
        '''Create and align widgets'''

        self.passwordout = StringVar()
        self.passwordout.set('- No password generated -')
        self.isnew = IntVar()

        ttitle = Label(master, text=versionStr, font=self.getFont(4))
        wisnew = Checkbutton(master, height=2, font=self.getFont(),
                text="This is a new password, that I have not used before",
                variable=self.isnew, command=self.OnNew)
        tlabel = Label(master, text='Label', font=self.getFont(2))
        tpasswordin1 = Label(master, text='Password', font=self.getFont(2))
        tpasswordin2 = Label(master, text='Password (again)',
                font=self.getFont(2))
        self.label = Entry(master, width=27, font=self.getFont())
        self.passwordin1 = Entry(master, width=27, font=self.getFont(),
                show="*")
        self.passwordin2 = Entry(master, width=27, font=self.getFont(),
                show="*", state=DISABLED)
        genbutton = Button(master, text="Generate password",
                font=self.getFont(), command=self.OnGen)
        clrbutton = Button(master, text="Clear fields", font=self.getFont(),
                command=self.OnClear)
        self.result = Entry(master, font=self.getFont(4),
                textvariable=self.passwordout, state="readonly", fg="black",
                readonlybackground="gray")

        # Layout widgets in a grid
        ttitle.grid(row=0, column=0, sticky=N+S+E+W, columnspan=3)
        wisnew.grid(row=1, column=0, sticky=N+S+E+W, columnspan=3)
        tlabel.grid(row=2, column=0, sticky=N+S+E+W)
        tpasswordin1.grid(row=2, column=1, sticky=N+S+E+W)
        tpasswordin2.grid(row=2, column=2, sticky=N+S+E+W)
        self.label.grid(row=3, column=0, sticky=N+S+E+W)
        self.passwordin1.grid(row=3, column=1, sticky=N+S+E+W)
        self.passwordin2.grid(row=3, column=2, sticky=N+S+E+W)
        genbutton.grid(row=4, column=0, sticky=N+S+E+W, columnspan=2)
        clrbutton.grid(row=4, column=2, sticky=N+S+E+W)
        self.result.grid(row=5, column=0, sticky=N+S+E+W, columnspan=3)

        # Initially, set focus on self.label
        self.label.focus_set()

def CLI(labels, pwTypoCheck=False, useCurses=True):

    while True:
        inp = getpass.getpass()
        if pwTypoCheck:
            inp2 = getpass.getpass("Again:")
            if inp == inp2:
                break;
            else:
                print "Passwords do not match. Please try again."
        else:
            break

    if useCurses:
        stdscr = curses.initscr()
        curses.noecho()
        curses.cbreak()

        stdscr.addstr(0, 0, versionStr + " - press 'q' to quit", curses.A_BOLD)
        stdscr.addstr(1, 0, "Your passphrase(s)", curses.A_BOLD)
        divlen = len(max(labels, key=len)) + 2
        j = 3
        for i in labels:
            stdscr.addstr(j,  0, i, curses.A_BOLD)
            stdscr.addstr(j, divlen, digest(i + inp), curses.A_REVERSE)
            j += 1
        del inp
        stdscr.refresh()

        while True:
            c = stdscr.getch()
            if c == ord('q'):
                break

        curses.nocbreak()
        curses.echo()
        curses.endwin()
    else:
        for i in labels:
            print "%25s %s" % (i, digest(i + inp))

def digest(message):
    '''Create and return secure hash of message

    A secure hash/message digest formed by hashing the `message` with
    the sha512 algorithm, encoding this hash with base64 and stripping
    it down to the first 30 characters.

    :Parameters:
        - `message`: The string from which to form the digest

    :Return:
        - The secure hash of `message`
    '''
    d = hashlib.sha512()
    d.update(message)
    shastr = d.hexdigest()

    # replace + and / with 4 and 9 respectively
    r = base64.b64encode(shastr, '49')
    r = r.replace('=','') # remove '=' if it's there
    r = r[:30]
    return str(r)

def usage():
    if os.name == 'nt':
        print "%s(nt) - http://babab.nl/p/dispass" % versionStr
        print
        print 'When DisPass is started without arguments, the graphical'
        print 'version will be started. To use the command line,'
        print 'submit one or more labels.'
        print
        print 'USAGE: dispass [options] [label] [label2] [label3] [...]'
        print
        print 'Options:'
        print '-c, --create    use if this passphrase is new (check input PW)'
        print '-h, --help      show this help and exit'
        print '-V, --version   show full version information and exit'
    else:
        print "%s(%s) - http://babab.nl/p/dispass" % (versionStr, os.name)
        print
        print "When DisPass is executed as 'gdispass' or 'dispass -g',"
        print 'the graphical version will be started.'
        print
        print 'USAGE: dispass [-co] label [label2] [label3] [...]'
        print '       dispass -g | -h | -V'
        print '       gdispass'
        print
        print 'Options:'
        print '-c, --create    use if this passphrase is new (check input PW)'
        print '-g, --gui       start guided graphical version of DisPass'
        print '-h, --help      show this help and exit'
        print '-o, --output    output passphrases to stdout (instead of the '
        print '                more secure way of displaying via curses)'
        print '-V, --version   show full version information and exit'

def main(argv):
    try:
        opts, args = getopt.getopt(argv[1:], "cghoV",
                ["create", "gui", "help", "output", "version"])
    except getopt.GetoptError, err:
        print str(err), "\n"
        usage()
        sys.exit(2)

    if args:
        labels = args
    else:
        labels = False

    pwTypoCheck = False
    for o, a in opts:
        if o in ("-g", "--gui"):
            GUI()
            return
        elif o in ("-c", "--create"):
            pwTypoCheck = True
        elif o in ("-o", "--output"):
            settings.useCurses = False
        elif o in ("-h", "--help"):
            usage()
            sys.exit()
        elif o in ("-V", "--version"):
            print versionStr, '-', __version_info__, 'running on', os.name
            sys.exit()
        else:
            assert False, "unhandled option"

    if labels:
        CLI(labels, pwTypoCheck, settings.useCurses)
    else:
        if os.name == 'nt':
            usage()
            print '-' * 78
            print 'Starting GUI...'
            GUI()
        else:
            usage()

if __name__ == '__main__':
    main(sys.argv)
